packages/core/src/shared/utilities/resourceCache.ts

```diff
@@ -0,0 +1,193 @@
+ */
+export interface GlobalStateSchema<V> {
+    resource: Resource<V>
+}
+
+const logger = getLogger('resourceCache')
+
+function now() {
+    return globals.clock.Date.now()
+}
+
+/**
+ * CacheResource utilizes VSCode global states API to cache resources which are expensive to get so that the result can be shared across multiple VSCode instances.
+ *  The first VSCode instance invoking #getResource will hold a lock and make the actual network call/FS read to pull the real response.
+ *  When the pull is done, the lock will be released and it then caches the result in the global states. Then the rest of instances can now acquire the lock 1 by 1 and read the resource from the cache.
+ *
+ * constructor:
+ *  @param key: global state key, which is used for globals.globalState#update, #tryGet etc.
+ *  @param expirationInMilli: cache expiration time in milli seconds
+ *  @param defaultValue: default value for the cache if the cache doesn't pre-exist in users' FS
+ *  @param waitUntilOption: waitUntil option for acquire lock
+ *
+ * methods:
+ *  @method resourceProvider: implementation needs to implement this method to obtain the latest resource either via network calls or FS read
+ *  @method getResource: obtain the resource from cache or pull the latest from the service if the cache either expires or doesn't exist
+ */
+export abstract class CachedResource<V> {
+    constructor(
+        private readonly key: globalKey,
+        private readonly expirationInMilli: number,
+        private readonly defaultValue: GlobalStateSchema<V>,
+        private readonly waitUntilOption: { timeout: number; interval: number; truthy: boolean }
+    ) {}
+
+    abstract resourceProvider(): Promise<V>
+
+    async getResource(): Promise<V> {
+        const cachedValue = await this.tryLoadResourceAndLock()
+        const resource = cachedValue?.resource
+
+        // If cache is still fresh, return cached result, otherwise pull latest from the service
+        if (cachedValue && resource && resource.result) {
+            const duration = now() - resource.timestamp
+            if (duration < this.expirationInMilli) {
+                logger.debug(
+                    `cache hit, duration(%sms) is less than expiration(%sms), returning cached value %s`,
+                    duration,
+                    this.expirationInMilli,
+                    this.key
+                )
+                // release the lock
+                await this.releaseLock(resource, cachedValue)
+                return resource.result
+            } else {
+                logger.debug(
+                    `cache is stale, duration(%sms) is older than expiration(%sms), pulling latest resource %s`,
+                    duration,
+                    this.expirationInMilli,
+                    this.key
+                )
+            }
+        } else {
+            logger.debug(`cache miss, pulling latest resource %s`, this.key)
```

Comment:
this one should be at least INFO level
```suggestion
            logger.info(`cache miss, pulling latest resource %s`, this.key)
```